src/coprocessor/dag/expr/builtin_time.rs

```diff
@@ -24,9 +24,26 @@ impl ScalarFunc {
         ctx: &mut EvalContext,
         row: &'a [Datum],
     ) -> Result<Option<Cow<'a, [u8]>>> {
-        let t = try_opt!(self.children[0].eval_time(ctx, row));
+        let t = match self.children[0]
+            .eval_time(ctx, row)
+            .and_then(|some_time| match some_time {
+                None => Err(Error::incorrect_datetime_value("None")),
+                Some(res) => Ok(res),
+            })
+            .map_err(|e| {
+                Error::handle_invalid_time_error(ctx, e)
+                    .or_else(Err)
+                    .and_then(|()| Ok(None))
+            }) {
```

Comment:
```rust
let t = match ... {
    Err(err) => return err,
    Ok(x) => x,
};
```

can be simplified to `let t = ...?;`. Notice the `?` operator.

Reply:
it is not easy to just use `?` instead of match.

```
the trait `std::convert::From<std::result::Result<std::option::Option<_>, coprocessor::codec::error::Error>>` is not implemented for `coprocessor::codec::error::Error
```

so I think this can use `handler_try!` like `try_opt!`.

Reply:
@breeswish  PTAL